### ISSUE TITLE
importing 4096 bit rsa keys from pkcs12 into Nitrokey HSM 2

### ISSUE THREAD

**Issue by Blub** [CardContact/scsh-scripts#5]
Using a Nitrokey HSM 2 I can import 2048 bit pkcs12 keys (generated with gpg -> gpgsm) via keymanager, but if I try the exact same procedure with a 4096 bit key I get:
```
Importing key and certificate...
GPError: Card (CARD_INVALID_SW/27264) - "Unexpected SW1/SW2=6A80 (Checking error: Incorrect parameter in the command data field) received" in /home/myuser/scsh3.15.359/scsh/sc-hsm/SmartCardHSM.js#1238
    at /home/myuser/scsh3.15.359/scsh/sc-hsm/SmartCardHSM.js#1238
    at /home/myuser/scsh3.15.359/scsh/sc-hsm/HSMKeyStore.js#300
    at /home/myuser/scsh3.15.359/keymanager/keymanager.js#1884
    at /home/myuser/scsh3.15.359/keymanager/keymanager.js#2038
```

Generating a 4096 bit key works, so the device at least supports them...

**Comment by CardContact** [comment]
3072 and 4096 bit RSA keys must be supplied in the PrivateExponent/Modulus format rather than in CRT format. I suspect that your key (like the 2048 bit key) is in CRT format.

**Comment by Blub** [comment]
I use `gpgsm --export-secret-key-p12` in both cases. Does scsh contain any helpers I could add to the import function to change the format? I'm not finding many options in the openssl cli tools about the format at least...

**Comment by CardContact** [comment]
Unfortunately there is currently no function in scsh to convert from CRT to PE/M. The math is explained in [1].

[1] https://crypto.stackexchange.com/questions/19407/calculate-the-rsa-private-exponent-from-the-crt-parameters

**Comment by Blub** [comment]
So judging by the code then, the `keyblob` sent in SmartCardHSM.js would have to have a different layout, which comes from `DKEK.js` which just concatenates CRT values. Assuming I get the right values, would that be the point where I'd have to build a different buffer? What order/layout would it need to have?

Edit: `openssl rsa -in key.pem -noout -text` actually shows the private modulus to copy/paste and turn into a byte array in js btw ;-)

**Comment by CardContact** [comment]
The format is defined in table 8 in the user manual, available in the CDN.

It's basically

8 byte random data
2 byte key size
Private Exponent (2 byte length and value)
Modulus  (2 byte length and value)
Public Exponent (2 byte length and value)

**Comment by Blub** [comment]
Still getting the same error after changing DKEK.js to write out the private exponent in place of DP1/DQ1/P/PQ/Q, leaving the key size, private exponent, modulus, public exponent in that order as above.

(`crypto.decrypt()` with a private `Key` created with the calculated private exponent successfully decrypts the data encrypted with the public key, so that part seems to be fine)

**Comment by CardContact** [comment]
You could try it the other way round: Generate a 4K key on the sc-hsm, export the key and verify that your block matches the block generated by the card.

**Comment by Blub** [comment]
Ah yes, apparently it uses a key type of `5` instead of `6` with bigger keys. Now it imported.
I used BN.js for the math btw. as while testing the data `ByteString.modInverse` gave me different results to the tools I used to compare, and they couldn't load the key with the resulting data. (Then again both were different from what `openssl rsa -text` shows (though that one is usable...))
With BN.js I just ended up doing `e.invm(n.sub(p).sub(q).addn(1))` for `d`.

```
diff --git a/scsh/sc-hsm/DKEK.js b/scsh/sc-hsm/DKEK.js
index 35fc4cf..2bfe31c 100644
--- a/scsh/sc-hsm/DKEK.js
+++ b/scsh/sc-hsm/DKEK.js
@@ -39,6 +39,7 @@ function DKEK(crypto) {

 exports.DKEK = DKEK;

+BN = require("util/bn").BN;


 /**
@@ -300,6 +301,31 @@ DKEK.prototype.encodeAESKey = function(key) {
 }


+const getComponentHex = function(key, component) {
+       return (''+key.getComponent(component)).replace(/\s/g, '');
+};
+
+const GetComponentBn = function(key, component) {
+       var hex = getComponentHex(key, component);
+       return new BN(hex, 16, 'be');
+};
+
+const BnToByteString = function(bn) {
+       return new ByteString(bn.toString(16), HEX);
+};
+
+const getPrivateExponentBn = function(key, pub) {
+       var n = GetComponentBn(pub, Key.MODULUS);
+       var e = GetComponentBn(pub, Key.EXPONENT);
+       var p = GetComponentBn(key, Key.CRT_P);
+       var q = GetComponentBn(key, Key.CRT_Q);
+       var phi = n.sub(p).sub(q).addn(1);
+       return e.invm(phi);
+};
+
+const getPrivateExponent = function(key, pub) {
+       return BnToByteString(getPrivateExponentBn(key, pub));
+};

 /**
  * Wrap RSA or ECC key
@@ -314,7 +340,12 @@ DKEK.prototype.encodeKey = function(pri, pub) {
        bb.append(this.getKCV());

        if (pub.getComponent(Key.MODULUS)) {
+               var bits = pub.getComponent(Key.MODULUS).length << 3;
+               if (bits > 2048) {
+                       bb.append(5);
+               } else {
                        bb.append(6);
+               }
                var algo = new ByteString("id-TA-RSA-v1-5-SHA-256", OID);
        } else {
                bb.append(12);
@@ -332,8 +363,15 @@ DKEK.prototype.encodeKey = function(pri, pub) {
        kb.append(this.crypto.generateRandom(8));

        if (pub.getComponent(Key.MODULUS)) {
-               kb.append(ByteString.valueOf(pub.getComponent(Key.MODULUS).length << 3, 2));
+               var bits = pub.getComponent(Key.MODULUS).length << 3;
+               kb.append(ByteString.valueOf(bits, 2));

+               if (bits > 2048) {
+                       print("Found key > 2048 bits, using different encoding");
+                       var c = getPrivateExponent(pri, pub);
+                       kb.append(ByteString.valueOf(c.length, 2));
+                       kb.append(c);
+               } else {
                        var c = pri.getComponent(Key.CRT_DP1);
                        kb.append(ByteString.valueOf(c.length, 2));
                        kb.append(c);
@@ -353,6 +391,7 @@ DKEK.prototype.encodeKey = function(pri, pub) {
                        var c = pri.getComponent(Key.CRT_Q);
                        kb.append(ByteString.valueOf(c.length, 2));
                        kb.append(c);
+               }

                var c = pub.getComponent(Key.MODULUS);
                kb.append(ByteString.valueOf(c.length, 2));
```

You should really just support regular imports via standard tooling. This dancing around java/javascript/weirdness is a PITA. I know you think generating keys on the hardware is better, but even better is leaving the choice to the users. Even hardware RNGs can be bad, or badly used[1].

[1] https://www.yubico.com/2017/10/infineon-rsa-key-generation-issue/

**Comment by Blub** [comment]
Seems to be working.

**Comment by CardContact** [comment]
Included in version 3.15.383 of the [Smart Card Shell](https://www.openscdp.org/scsh3/download.html).
